pkg/authn/keychain.go

```diff
@@ -35,7 +35,19 @@ type Keychain interface {
 
 // defaultKeychain implements Keychain with the semantics of the standard Docker
 // credential keychain.
-type defaultKeychain struct{}
+type defaultKeychain struct {
+	configDir string
+}
+
+// SetDefaultKeychainConfigDir sets the client config directory to specified location.
+func SetDefaultKeychainConfigDir(p string) error {
+	dk, ok := DefaultKeychain.(*defaultKeychain)
```

Comment:
What about making this function called on the *defaultKeychain similar to the Resolve function below? You just set the configDir to p here. And when you call it, you call it on the exported DefaultKeychain instance auth.DefaultKeychain, which is more similar to how pull.go is currently consuming auth.DefaultKeychain when invoking Resolve(), or the python implementation, which seems making a little bit more sense to me.

If you are not planning to implement consuming the new configDir attribute in configDir() or Resolve() function, maybe add a TODO to keep track. I would probably do that together in this PR as it's small and making this feature complete.

Reply:
This wouldn't work as is because "DefaultKeychain" is of type "Keychain" which is an interface that doesn't have the "SetConfigDir" function. If the function were added to the interface, every implementation of the interface would need to add this function. This would be an API breaking update to go-containerregistry as "Keychain" is part of the public API.